pinot-server/src/main/java/org/apache/pinot/server/starter/helix/SegmentOnlineOfflineStateModelFactory.java

```diff
@@ -151,12 +156,14 @@ public void onBecomeDroppedFromConsuming(Message message, NotificationContext co
       _logger.info("SegmentOnlineOfflineStateModel.onBecomeDroppedFromConsuming() : " + message);
       String realtimeTableName = message.getResourceName();
       String segmentName = message.getPartitionName();
-      TableDataManager tableDataManager = _instanceDataManager.getTableDataManager(realtimeTableName);
-      Preconditions.checkState(tableDataManager != null, "Failed to find table: %s", realtimeTableName);
-      tableDataManager.onConsumingToDropped(segmentName);
       try {
         _instanceDataManager.offloadSegment(realtimeTableName, segmentName);
         _instanceDataManager.deleteSegment(realtimeTableName, segmentName);
+        // At this point ingestion thread should be stopped, we notify tableManager so it can remove its resources
+        // safely
+        TableDataManager tableDataManager = _instanceDataManager.getTableDataManager(realtimeTableName);
+        Preconditions.checkState(tableDataManager != null, "Failed to find table: %s", realtimeTableName);
+        tableDataManager.onConsumingToDropped(segmentName);
```

Comment:
I feel that we should also change the method name `onConsumingToDropped` to `onConsumingToDroppedOrOffline`. Otherwise, the name is mis-leading